sys-fs/zfs-kmod: clean up old modules in live ebuild

gyakovlev opened the pull request:
untested, just opening for visibility.

not sure if it handles compressed modules, probably not.

thesamesam commented on sys-fs/zfs-kmod/zfs-kmod-9999.ebuild:
```diff
@@ -200,4 +200,27 @@ pkg_postinst() {
+		kopath="${EROOT}/lib/modules/${KV_FULL}/extra/${ko}"
+		if [[ -f "${kopath}" ]]; then
+			ewarn "Found previously installed zfs module ${ko} for current kernel ${KV_FULL}"
+			rm -v "${kopath}" || die
```
Luckily, the list of suffixes is so small that we can just list them, I think (gz, lz4, xz)? Would have to check.

thesamesam reviewed (commented):
Not tried it either but this looks right & thank you for doing it

gyakovlev commented:
```
 * Removing sys-fs/zfs-kmod-2.1.7 from moduledb.
>>> Original instance of package unmerged safely.
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/avl/zavl.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/avl'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/icp/icp.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/icp'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/lua/zlua.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/lua'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/nvpair/znvpair.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/nvpair'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/spl/spl.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/spl'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/unicode/zunicode.ko' 
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/unicode'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zcommon/zcommon.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zcommon'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zfs/zfs.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zfs'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zstd/zzstd.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zstd'
 * Found obsolete zfs module for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zstd/zzstd.ko.gz'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zstd'
 * Updating module dependencies for 5.15.80-talos64 ...                                                                  [ ok ]
 * Adding module to moduledb.                                   
 * Building initramfs via dracut ...
...

tree /lib/modules/5.15.80-talos64/extra
/lib/modules/5.15.80-talos64/extra
├── spl.ko
└── zfs.ko

0 directories, 2 files
```

gyakovlev commented:
also I think we need the check to file only if new file exists as well, in case someone pins 9999 to something older.

gyakovlev commented:
```Hack
* Removing sys-fs/zfs-kmod-2.1.7 from moduledb.
>>> Original instance of package unmerged safely.
 * Found obsolete zfs module avl/zavl.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/avl/zavl.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/avl'
 * Found obsolete zfs module icp/icp.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/icp/icp.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/icp'
 * Found obsolete zfs module lua/zlua.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/lua/zlua.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/lua'
 * Found obsolete zfs module nvpair/znvpair.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/nvpair/znvpair.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/nvpair'
 * Found obsolete zfs module spl/spl.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/spl/spl.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/spl'
 * Found obsolete zfs module unicode/zunicode.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/unicode/zunicode.ko' 
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/unicode'
 * Found obsolete zfs module zcommon/zcommon.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zcommon/zcommon.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zcommon'
 * Found obsolete zfs module zfs/zfs.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zfs/zfs.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zfs'
 * Found obsolete zfs module zstd/zzstd.ko for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zstd/zzstd.ko'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zstd'
 * Found obsolete zfs module zstd/zzstd.ko.gz for current kernel 5.15.80-talos64, removing.
removed '/lib/modules/5.15.80-talos64/extra/zstd/zzstd.ko.gz'
rmdir: removing directory, '/lib/modules/5.15.80-talos64/extra/zstd'
 * Updating module dependencies for 5.15.80-talos64 ...                                                                  [ ok ]
 ```

gyakovlev commented:
ok I think I'm  mostly happy with it now and no edge cases come to mind.
I hate removing files in such way, but otherwise depmod and initrd might get confused unfortunately.

thesamesam commented on sys-fs/zfs-kmod/zfs-kmod-9999.ebuild:
```diff
@@ -174,7 +174,44 @@ src_install() {
+
 pkg_postinst() {
+	# check for old module layout before doing anything else.
+	# only attempt layout cleanup if new .ko location is used.
```
nice sanity check

gyakovlev replied:
After sleeping on it, needs an extension though… need to check compressed too here.

thesamesam reviewed (approved):
Thanks, this is done well
